Have AsyncAudioDecoder use NativePromise

jyavenard opened the pull request:
#### 4667cdc966a45d336d34283e302d1e1a77a2420a
<pre>
Have AsyncAudioDecoder use NativePromise
<a href="https://bugs.webkit.org/show_bug.cgi?id=261601">https://bugs.webkit.org/show_bug.cgi?id=261601</a>
rdar://115549443

Reviewed by Youenn Fablet.

The purpose of this change is mostly to show how some threading work can
be greatly simplified with the new NativePromise object.

No change in observable behaviour.

* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
(WebCore::AsyncAudioDecoder::AsyncAudioDecoder):
(WebCore::AsyncAudioDecoder::decodeAsync):
(WebCore::AsyncAudioDecoder::~AsyncAudioDecoder): Deleted.
(WebCore::AsyncAudioDecoder::runLoop): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::DecodingTask): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::decode): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::notifyComplete): Deleted.
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):

Canonical link: <a href="https://commits.webkit.org/268236@main">https://commits.webkit.org/268236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e50b89c55089a10d47adc8a797882dd8d56dfa8f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18968 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20836 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17734 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19438 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19510 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19302 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21725 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16510 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17285 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23693 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17550 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17457 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21634 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18036 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15295 "1 flakes 45 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17118 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4553 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17870 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
```diff
@@ -29,86 +29,27 @@
 }
 
-AsyncAudioDecoder::~AsyncAudioDecoder()
+Ref<DecodingTaskPromise> AsyncAudioDecoder::decodeAsync(Ref<ArrayBuffer>&& audioData, float sampleRate)
```
Great simplification in general.

youennf commented on Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
```diff
@@ -29,86 +29,27 @@
+    return WTF::invokeAsync(m_runLoop, __func__, [audioData = WTFMove(audioData), sampleRate] {
+        auto audioBuffer = AudioBuffer::createFromAudioFileData(audioData->data(), audioData->byteLength(), false, sampleRate);
+        if (!audioBuffer)
+            return DecodingTaskPromise::createAndReject(EncodingError, __func__);
```
At some point, we will want to pass an Exception object, crossThreadyCopy() plus ensuring that move trigger zero copies should make this possible I think.

jyavenard replied:
I don't see the need to be able to allocate an Exception object outside of the main thread and pass-it back.
There's zero need for it, it adds unnecessary complexity.

Just like how WTFString can't be used with WorkQueue, only RunLoops. NativePromise doesn't have magical capabilities that somehow could bypass the limitations of the objects it's dealing with.

youennf commented on Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
```diff
@@ -29,86 +29,27 @@
+        auto audioBuffer = AudioBuffer::createFromAudioFileData(audioData->data(), audioData->byteLength(), false, sampleRate);
+        if (!audioBuffer)
+            return DecodingTaskPromise::createAndReject(EncodingError, __func__);
+        return DecodingTaskPromise::createAndResolve(audioBuffer.releaseNonNull(), __func__);
```
Given AudioBuffer is RefCounted<>, we need to be extra sure about the thread safety model here.
The promise documentation says:
`When IsExclusive is true, move semantics are used when passing arguments whenever possible`.

The `whenever possible` is a bit worrying, the code should ensure thread safety no matter whether the value is movable-only like unique_ptr or both movable and copyable like RefPtr.
We should have clear guidelines on whether/how we can use RefCounted objects.

For instance, resolving a promise with a RefCounted object that has a ref count of more than 1 is potentially problematic given caller and callee might not know whether they work on the same thread.

jyavenard replied:
I updated the documentation in the NativePromise PR. Move will always be used. If the object isn't movable, then it won't compile. Simple as that.
From the documentation.
```
 * When IsExclusive is true:
 *  - The NativePromise performs run-time assertions that there is at most one call to either then(...) or chainTo(...).
 *  - Move semantics are used when passing arguments
 *  - The resolved or rejected object will be deleted on the target thread.
 *  - The ResolveValueType and RejectValueType must be moveable (e.g have a move constructor). Compilation will fail otherwise.
 * Otherwise:
 * - values are passed to the resolve/reject callbacks through either const references or pointers.
 * - the resolve or reject object will be deleted on the last SerialFunctionDispatcher that got used.
```
The entire chain of function parameters when IsExclusive is true is ResolveValueType&&

There's of course the issue if the promise isn't then()ed or whenSettled() then the scope of the value is the same as the promise. But that isn't the case here.

In this particular case, it is obvious what is going on, for the simple case that EWS shows no crash.
AudioBuffer isn't a DOM object, it has a non-thread-safe refcount and it is moved. If there was, RefCounted<AudioBuffer> would have asserted when it goes out of scope on the main thread.

Also, the flow of this AudioBuffer from decoder runLoop back to the main thread is identical to before. Why would 
consider it a problem all of a sudden?
Code before was doing exactly that: allocate on the decoder's runLoop, move it to the main thread and return it to the caller.

> For instance, resolving a promise with a RefCounted object that has a ref count of more than 1 is potentially problematic given caller and callee might not know whether they work on the same thread.

While NativePromise greatly simplify the use and add lots of safety layers, you would expect a WebKit programmer to have some clue on what they are doing when dealing with multi-threading and they decide to share a RefPtr<T> across multiple-threads and use it at the same time !

This change is all about using NativePromise, not arguing again on how NativePromise is written. I had hoped that had been settled by now :(

youennf commented on Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
```diff
@@ -298,16 +299,18 @@ void BaseAudioContext::decodeAudioData(Ref<ArrayBuffer>&& audioData, RefPtr<Audi
-            auto audioBuffer = result.releaseReturnValue();
+            auto audioBuffer = WTFMove(result.value());
             if (promise)
                 promise.value()->resolve<IDLInterface<AudioBuffer>>(audioBuffer.get());
```
Could we use resolveWithNewlyCreated instead and move audioBuffer?

jyavenard replied:
That seems out of scope tbh.
The aim of this PR was to replace one existing code with an identical one using NativePromise.

jyavenard replied:
Actually, I don't think we can.
audioBuffer is returned to both the promise and `successCallback` you can't move it twice.

youennf commented on Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
```diff
@@ -298,16 +299,18 @@ void BaseAudioContext::decodeAudioData(Ref<ArrayBuffer>&& audioData, RefPtr<Audi
-    m_audioDecoder->decodeAsync(WTFMove(audioData), sampleRate(), [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise)](ExceptionOr<Ref<AudioBuffer>>&& result) mutable {
+    // We don't move audioData as we want to delete it on the main thread upon completion of the DecodingTaskPromise.
+    auto p = m_audioDecoder->decodeAsync(Ref { audioData }, sampleRate());
+    p->then(RunLoop::main(), __func__, [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise), audioData = WTFMove(audioData)] (DecodingTaskPromise::Result&& result) mutable {
```
Why capturing audioData in the then callback?
Can we just keep passing/moving it to decodeAsync?

jyavenard replied:
>Why capturing audioData in the then callback?

I explained the why in the comment above.
audioData is a JS array buffer, it's main thread only.

We want it to be deleted on the main thread. If we were to move it do the decoder thread, it would be deleted on that thread which is no good.
By capturing in the callback object, it will be deleted on the main thread once decoding is done (or has failed)

youennf replied:
Isn't it the case that, with the current patch, we will unref audioData in the audioDecoder WorkQueue, as well as potentially refing/unrefing it in main thread in parallel?
Shouldn't we instead ASSERT that `decodeAsync()` is called in main thread and use callOnMainThread within decodeAysnc() to move back `audioData` to main thread?

youennf reviewed (approved):
LGTM.
It would be great if, in the future, we could return as DecodingTaskPromise::Result an ExceptionOr<Ref<AudioBuffer>>.
Then we could do something like promise.value()->settle(result).
Not as good here since it is mixing callbacks and promise handling but still...

jyavenard committed:
Have AsyncAudioDecoder use NativePromise
https://bugs.webkit.org/show_bug.cgi?id=261601
rdar://115549443

Reviewed by Youenn Fablet.

The purpose of this change is mostly to show how some threading work can
be greatly simplified with the new NativePromise object.

No change in observable behaviour.

* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.cpp:
(WebCore::AsyncAudioDecoder::AsyncAudioDecoder):
(WebCore::AsyncAudioDecoder::decodeAsync):
(WebCore::AsyncAudioDecoder::~AsyncAudioDecoder): Deleted.
(WebCore::AsyncAudioDecoder::runLoop): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::DecodingTask): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::decode): Deleted.
(WebCore::AsyncAudioDecoder::DecodingTask::notifyComplete): Deleted.
* Source/WebCore/Modules/webaudio/AsyncAudioDecoder.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::decodeAudioData):

Canonical link: https://commits.webkit.org/268236@main
